consensus/src/processes/pruning_proof/mod.rs

```diff
@@ -614,45 +791,233 @@ impl PruningProofManager {
+            .parents_by_level
+            .iter()
+            .enumerate()
+            .skip(1)
+            .find_map(|(level, parents)| {
+                if BlockHashSet::from_iter(parents.iter().copied()) == direct_parents {
+                    None
+                } else {
+                    Some((level - 1) as BlockLevel)
+                }
+            })
+            .unwrap_or(self.max_block_level)
+    }
+
+    fn estimated_blue_depth_at_level_0(&self, level: BlockLevel, level_depth: u64, current_dag_level: BlockLevel) -> u64 {
+        level_depth.checked_shl(level.saturating_sub(current_dag_level) as u32).unwrap_or(level_depth)
+    }
+
+    /// selected parent at level = the parent of the header at the level
+    /// with the highest blue_work (using score as work in this case)
+    fn find_selected_parent_header_at_level(
+        &self,
+        header: &Header,
+        level: BlockLevel,
+    ) -> PruningProofManagerInternalResult<Arc<Header>> {
+        // Parents manager parents_at_level may return parents that aren't in relations_service, so it's important
+        // to filter to include only parents that are in relations_service.
+        let sp = self
+            .parents_manager
+            .parents_at_level(header, level)
+            .iter()
+            .copied()
+            .filter(|p| self.level_relations_services[level as usize].has(*p).unwrap())
+            .filter_map(|p| self.headers_store.get_header(p).unwrap_option().map(|h| SortableBlock::new(p, h.blue_work)))
+            .max()
+            .ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof("no parents with header".to_string()))?;
+        Ok(self.headers_store.get_header(sp.hash).expect("unwrapped above"))
+    }
+
+    /// Find a sufficient root at a given level by going through the headers store and looking
+    /// for a deep enough level block
+    /// For each root candidate, fill in the ghostdag data to see if it actually is deep enough.
+    /// If the root is deep enough, it will satisfy these conditions
+    /// 1. block at depth 2m at this level ∈ Future(root)
+    /// 2. block at depth m at the next level ∈ Future(root)
+    ///
+    /// Returns: the filled ghostdag store from root to tip, the selected tip and the root
+    fn find_sufficient_root(
+        &self,
+        pp_header: &HeaderWithBlockLevel,
+        level: BlockLevel,
+        current_dag_level: BlockLevel,
+        required_block: Option<Hash>,
+        temp_db: Arc<DB>,
+    ) -> PruningProofManagerInternalResult<(Arc<DbGhostdagStore>, Hash, Hash)> {
+        // Step 1: Determine which selected tip to use
+        let selected_tip = if pp_header.block_level >= level {
+            pp_header.header.hash
+        } else {
+            self.find_selected_parent_header_at_level(&pp_header.header, level)?.hash
+        };
+
+        let cache_policy = CachePolicy::Count(2 * self.pruning_proof_m as usize); // TODO: We can probably reduce cache size
```

Comment:
I think this TODO can be removed. 2M is what most levels need, and it's not a large number in the first place even if not used by some higher levels